packages/@lwc/module-resolver/src/index.ts

```diff
@@ -4,117 +4,147 @@
+): RegistryEntry | undefined {
+    const { dir } = moduleRecord;
+    const absModuleDir = path.isAbsolute(dir) ? dir : path.join(opts.rootDir, dir);
+    const [ns, name] = specifier.split('/');
 
-export type ModuleRecord = string | ModuleRecordObject;
-export interface LwcConfig {
-    modules: ModuleRecord[];
+    if (ns && name) {
+        const moduleDir = path.join(absModuleDir, ns, name);
+        const entry = getModuleEntry(moduleDir, name);
+        if (entry) {
+            return createRegistryEntry(entry, specifier, opts);
+        }
+    }
 }
 
-function resolveModulesFromDir(modulesDir: string): RegistryEntry[] {
-    const namespaces = fs.readdirSync(modulesDir);
-    const resolvedModules: RegistryEntry[] = [];
-    namespaces.forEach(ns => {
-        if (ns[0] !== '.' && fs.lstatSync(path.join(modulesDir, ns)).isDirectory()) {
-            const namespacedModuleDir = path.join(modulesDir, ns);
-            const modules = fs.readdirSync(namespacedModuleDir);
-            modules.forEach(moduleName => {
-                const moduleDir = path.join(namespacedModuleDir, moduleName);
-                const entry = getModuleEntry(moduleDir, moduleName);
-                if (entry) {
-                    const specifier = `${ns}/${moduleName}`;
-                    resolvedModules.push({ entry, specifier });
+function resolveModuleFromNpm(
+    specifier: string,
+    npmModuleRecord: NpmModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { npm, map: aliasMapping } = npmModuleRecord;
+    const pkgJsonPath = require.resolve(`${npm}/package.json`, { paths: [opts.rootDir] });
+    const packageDir = path.dirname(pkgJsonPath);
+    const lwcConfig = getLwcConfig(packageDir);
+    const modules = lwcConfig && lwcConfig.modules;
+
+    if (modules) {
+        validateNpmConfig(lwcConfig);
+        const exposedModules = lwcConfig.expose || [];
+        if (exposedModules.includes(specifier)) {
+            for (const moduleRecord of modules) {
+                if (!isNpmModuleRecord(moduleRecord)) {
+                    const registryEntry = resolveModuleRecordType(specifier, moduleRecord, {
+                        rootDir: packageDir,
+                    });
+                    if (registryEntry) {
+                        if (aliasMapping && aliasMapping[specifier]) {
+                            registryEntry.specifier = aliasMapping[specifier];
+                        }
+                        return registryEntry;
+                    }
                 }
-            });
+            }
+
+            throw new Error(`Unable to find ${specifier} under package ${npmModuleRecord.npm}`);
         }
-    });
+    }
+}
 
-    return resolvedModules;
+function resolveModuleRecordType(
+    specifier: string,
+    moduleRecord: ModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { rootDir } = opts;
+    validateModuleRecord(moduleRecord);
```

Comment:
I don't think at this point this method can throw since we are doing a sanitization of the module records. We should move the validation to the `resolveModule` and iterate over all the `ModuleRecord`s and validate them first.